In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv('wc-wo-outliers.csv')

In [3]:
y = df['results']
X = df.drop(columns=['results'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=2)

In [4]:
regr = MLPRegressor(random_state=1, hidden_layer_sizes=(1580), learning_rate='invscaling',
                    solver='sgd', power_t=.49, max_iter=600).fit(X_train, y_train)


/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (600) reached and the optimization hasn't converged yet.
  warnings.warn(


In [5]:
regr.predict(X_test[:2])

array([2.69096029, 3.07047391])

In [6]:
regr.score(X_test, y_test)

-1.1588778924721903

In [7]:
y_pred = regr.predict(X_test)

In [8]:
y_train_pred = regr.predict(X_train)

In [19]:
print('Neural Network Regression')
print('r2 score: '+str(r2_score(y_test, y_pred)))
print('RMSE : '+str(np.sqrt(mean_squared_error(y_test, y_pred))))
print('MSE: '+str(mean_squared_error(y_test, y_pred)))
print('MAE: '+str(mean_absolute_error(y_test, y_pred)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, y_pred)))

# print('----------------')
# print('r2 score: '+str(r2_score(y_train, y_train_pred)))
# print('RMSE : '+str(np.sqrt(mean_squared_error(y_train, y_train_pred))))
# print('MSE: '+str(mean_squared_error(y_train, y_train_pred)))
# print('MAE: '+str(mean_absolute_error(y_train, y_train_pred)))
# print('MAPE: '+str(mean_absolute_percentage_error(y_train, y_train_pred)))

Neural Network Regression
r2 score: -1.1588778924721903
RMSE : 12.855249923451426
MSE: 165.25745059439788
MAE: 9.746986045650164
MAPE: 0.7569210486642853


In [10]:
df.shape

(184, 17)

In [11]:
# mape on test .766 @ (182, 17)

## Neural Network Regression
- Neural Network Regression
- r2 score: -1.1588778924721903
- RMSE : 12.855249923451426
- MSE: 165.25745059439788
- MAE: 9.746986045650164
- MAPE: 0.7569210486642853


In [12]:
data = {'y_test': y_test, 'y_pred': y_pred}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)

In [13]:
new_df = new_df.sort_values(by='diff', ascending=True)

In [14]:
new_df['y_pred'] = round(new_df['y_pred'], 1)

In [15]:
new_df

y_test  y_pred       diff
29        3     3.0   0.006041
109       3     2.9   0.134331
25        2     2.8   0.837548
23        4     3.1   0.919608
35        4     2.9   1.084592
10        4     2.8   1.245381
6         1     2.9   1.884718
3         1     2.9   1.904248
53        1     2.9   1.911591
2         5     2.9   2.106268
155       5     2.7   2.284170
119       5     2.7   2.301281
114       6     2.9   3.063885
161       6     2.8   3.162073
99        7     2.8   4.163781
154       9     2.7   6.281920
138       9     2.7   6.329899
54       10     2.7   7.265946
173      10     2.6   7.431096
41       12     2.9   9.091644
85       12     2.8   9.222308
45       12     2.8   9.232545
13       13     2.9  10.140900
164      14     2.9  11.137595
152      16     2.7  13.267494
172      17     2.8  14.176570
89       17     2.7  14.280481
12       18     2.8  15.204950
57       21     2.9  18.075853
79       22     3.1  18.929526
125      22     2.7  19.271211
167      23     2.9  20.146661
64       25     2.9  22.108281
122      26     2.9  23.098436
160      27     2.8  24.209632
48       30     2.7  27.309040
127      30     2.6  27.416981

In [16]:
new_df['diff'].sum() / len(new_df)

9.746986045650164

In [17]:
new_df['diff'].mean()

9.746986045650164

In [18]:
new_df['diff'].median()

7.43109589709514